### DISCUSSION TITLE
Initial RFC: Scope and Goals

### DISCUSSION THREAD

**Discussion by markerikson** [markerikson/react-community-tools-practices-cheatsheet#1] [Ideas]
I'll just go ahead and paste the tweets I pasted into the README as a starting point:

> it feels like there maybe needs to be more of a community effort to build up some kind of a centralized "best practices" site, perhaps similar to the React+TS CheatSheet ( https://github.com/typescript-cheatsheets/react )
> what if the community collaborated on putting together a centralized list of what the major options are in different categories, and when it might make sense to use them?  
> Obviously this could easily devolve into "NO, YOU SHOULD BE USING TOOL X INSTEAD OF TOOL Y", or "WHY ISN'T MY LIB WITH 27 STARS LISTED?".  There'd need to be both curation and friendly cooperation from experts within the community in this.
> Example, state management. The main options by usage % are plain React state, Redux, MobX, XState. Those overlap with data fetching libs like Apollo, React Query, and SWR. What if each lib submitted a page saying "here's our tool, what it solves, and when you might use it"  
> Similarly, styling. Major options are inline styles, plain CSS, CSS Modules, and CSS-in-JS approaches. Have some pages that detail what those options are, pros/cons, and why you might consider each approach. Not flat-out saying "use X or Y". Just "here's options and use cases"
> One more: how do you pick between CRA, Next, Gatsby, or something like a Snowpack or Vite, to build a React app? When does it make sense to use any of those? Having a list of the major options and tradeoffs involved would be valuable.  
> Stackshare and Slant are _sorta_ along the lines of what I'm describing, although I'm picturing something much more text-based and involving descriptions, examples, and use cases.
> I'm not saying any of this would be easy, and there'd be a lot of wrangling over how to format and present this info. also, not all would be "competing libs". like, `fetch` vs `axios` vs `redaxios` vs your own fetch wrapper

What I'm picturing here is somewhere in between a "list of top libraries", a "list of commonly used techniques and patterns", and a "examples for some best real-world best practices and advice to augment the React core docs".  The React team justifiably has avoided offering suggestions that extend beyond the scope of React itself, both to avoid appearance of favoritism, and because they want to let the ecosystem thrive.  I'd like to put together a community resource that kind of fills in the areas that the core React docs don't cover, offers info on techniques and practices for building real-world apps, and lists the pros/cons/use cases for areas that suffer from fragmentation and "analysis paralysis" ( https://twitter.com/floydophone/status/1365833622700220419 ).

So, this is me sorta putting my money where my mouth is, and creating a repo to start this. 

I guess the place to start is discussing:

- What would the scope of this be?
- What specific problems are we trying to solve with putting together a site like this?
- What content should be in here?
- How should that information be formatted?
- How do we curate this and prevent it from becoming yet another "Awesome List" with 5000 random links, and keep it useful, while also not restricting it to "here's the top 3 libs in a category and therefore only they get promoted more"?
- Who's the target audience?  Newbies, experienced devs, both at once?  (thanks to https://twitter.com/tenebrity/status/1365900999404236803 )

Additional thoughts:

- I've heard the "Two Scoops of Django" book cited as a good example of a "community best practices" document ( ref: https://twitter.com/acemarke/status/896415376296341504)
- Would be nice to see some of the "why use tool X" descriptions also say "here's when to maybe consider tool Y instead" ( https://twitter.com/stalefries/status/1365909893698392068 ), like having the Redux entry say "use MobX if you prefer OOP over functional or want more 'magic' in your state update handling"

**Comment by cseas** [comment]
Having setup a greenfield project recently, I can confirm setting up a React project for an enterprise requirement is a **nightmare**. 

There are dozens of tools in the React ecosystem to do pretty much everything and the community has divided opinions on each of them even if they solve different problems altogether.

The [part of the React docs about toolchains](https://reactjs.org/docs/create-a-new-react-app.html) is supposed to help with this but let's admit it, most React devs have never even seen it. 

The fact I dislike most about the React docs is that they fail to mention explicitly the pitfalls of create-react-app, a major one being that implementing any server-side rendering or adding things outside of that one root div with it is a sparsely documented herculean feat, unless we use third-party solutions like react-helmet, which haven't been maintained in years.

And this is a problem with many other aspects of a React app. There is so much dependency on third-party solutions to efficiently achieve anything but the official docs have no opinion on how to do anything in a large scale app. For comparison, Express and Redux are supposed to be unopinionated as well but they both have a best practices section to help beginners avoid making really well known common mistakes.

This initiative can help solve that. It can help people understand react-query isn't a state management tool, it can help people understand what XState is for, what ReduxToolkit does, what Apollo does different than SWR, what server rendering is & why most corporate apps probably need it in some capacity, how to just get some data from backend without dumping a temporary copy of database in your redux store and hundreds of other common things almost everyone is confused about.

**Reply by benjaminsuch** [reply]
I think you mix responsibilities here.

> The fact I dislike most about the React docs is that they fail to mention explicitly the pitfalls of create-react-app, a major one being that implementing any server-side rendering or adding things outside of that one root div with it is a sparsely documented herculean feat, unless we use third-party solutions like react-helmet, which haven't been maintained in years.

create-react-app is, as far as I know, not an official package of the React team and has been created to get people started. It was never supposed to be an all-in-one-deluxe-one-click-solution which solves all of your problems.

> And this is a problem with many other aspects of a React app. There is so much dependency on third-party solutions to efficiently achieve anything but the official docs have no opinion on how to do anything in a large scale app.

That is not a problem the React team has to be responsible for. React is an UI library. All the third party dependencies are a consequence of your app's requirements. You want to fetch data, state management, use SVG and CSS, SASS or CSS Modules etc. But that is out of the scope of React. If React's features do not meet your requirements, hey, you can still  write your own implementation and integrate it into your React app.

I think the React team did pretty well writing down best practices which only regards React and it's features.

At this point I think it would be a good idea for this initiative, to make people realize what React is and what not.

**Reply by markerikson** [reply]
FWIW, CRA is effectively an official React team project.  Dan and Christopher started it, and while the current maintainers aren't part of the React core team, it's still inside the Facebook Github org.

But yes, the React team has deliberately tried to limit their suggestions to using React itself, for a variety of reasons.

**Reply by cseas** [reply]
> At this point I think it would be a good idea for this initiative, to make people realize what React is and what not.

This can be a great point to focus on as well. I've lost count of how many devs I've met who thought React was synonymous with create-react-app. It's only natural that something they recommend in their docs tutorial will be assumed as official by most readers even if it's technically not.

Many debates about "React vs Angular" are really just "CRA vs Angular". Gatsby and Nextjs are so far hidden in the official docs, I've had to answer dozens of times the question: "Why're you using Nextjs, why not React?"

Quoting the docs, they do mention "[CRA is the best way to start building a new single-page application in React](https://reactjs.org/docs/create-a-new-react-app.html#create-react-app)". I'd say it's the simplest, but calling it best is highly subjective and will convince everyone that CRA is meant to be the one-stop solution for most problems, which it really isn't. Also, maybe I'm reading it wrong, but that statement in the docs is a strong opinion so the docs aren't completely unopinionated.

**Comment by 44px** [comment]
> - What would the scope of this be?
> - What specific problems are we trying to solve with putting together a site like this?

There is no good unopinionated guidance on libraries, so such site definitely can fill the gap.

I think this can be helpful for both newcomers and experienced developers. Use cases:
1. Starting new project. Even if we have prior experience, we don’t start new projects often, so some overview can be helpful here. Let's say I have Redux and styled-components on my current project and it serves me well. But there are some new practices gained popularity in last years I may consider for new project now:
    - useState/useReducer hooks for state + [unstated-next](https://www.npmjs.com/package/unstated-next) for sharing "global" state + [React Query](https://react-query.tanstack.com/) for caching server state
    - "utility CSS" solutions like Tailwind/[twin.macro](https://github.com/ben-rogerson/twin.macro)

    As a side note: maybe this site should document different approaches, but not specific libraries (compare approaches instead of libraries, use libraries only as implementation examples for every approach).

2. Picking library for existing project, evaluating solutions (like, “let’s replace SCSS with CSS-in-JS solution”)

3. Starting with React. This site may serve as an ecosystem overview for people having experience with other frameworks

> What content should be in here?

It would be good to start with some overview laying out foundation for further guidelines:
- What’s React used for?
  – "Classic" Single Page Apps (client rendering); Server-rendered apps, e-commerce, content sites; (Mostly) static sites; Mobile apps and so on... Maybe it should link to / embed ["wrapping up" section of this article on rendering](https://developers.google.com/web/updates/2019/02/rendering-on-the-web#wrapup)

- How should I pick solution?
  – Know what you are building, consider app and team size and teammates experience...

Then it can continue with guidelines for specific problems:
- Picking build tool, framework, boilerplate
- TypeScript or not?
- Routing
- State management
- Interacting with server
- Styling
- UI kits

> How should that information be formatted?

I like how @leerob structures his posts with comparisons:
- [How Should I Style My React Application?](https://leerob.io/blog/css-with-react)
- [Next.js vs. Gatsby vs. Create React App](https://leerob.io/blog/nextjs-gatsby-create-react-app)

Also, some libraries have own comparison pages ([Chakra UI](https://chakra-ui.com/docs/comparison), [React Query](https://react-query.tanstack.com/comparison)) so maybe this site should link to them if it would compare libraries

**Reply by markerikson** [reply]
**YES**!  I think @leerob 's posts there are almost _exactly_ the kind of material I'm picturing here!

- A summary of what the available options in this space are
- Brief descriptions of why you might consider this particular option
- Pros/cons/use cases for when this option works well

**Reply by alexandrudanpop** [reply]
Also wrote an [CRA vs Next.js vs Gatsby comparison](https://dev.to/alexandrudanpop/react-applications-when-to-pick-gatsby-or-next-js-or-create-react-app-50l1) that is less biased towards Next.js. Don't get me wrong, I think `Next.js` is awesome but there are valid use cases for the other two.

**Comment by benjaminsuch** [comment]
> What specific problems are we trying to solve with putting together a site like this?

I have the impression that people in general seek guidance of how to do stuff correctly, because the ecosystem and features of React can be quite overwhelming. If you don't have a good idea of data fetching, state management, data structures or how build-systems work in todays JavaScript world it's hard to figure out when to use X and when Y?

Example: Should I use `swr` or `react-query`? When and why?

But we have to be also careful to not go into "taste-topics" like: Should I use SASS with CSS Modules or CSS-in-JS? Styled Components? What should I pick for my giga-mega-highly-complex-and-special-app which contains at least millions of files?

I think it makes sense to make categories, something like:

 - **Common mistakes**
 - **Glossary**
   Which explains terms, for example what SSR is and why is it good and when does it matter for me. 
- **Use-Cases**
  A list of website-types, what a useful Tech-Stack would like to build them, what's special and important for these kind of websites and what pitfalls/mistakes to look out for.
  - Blogs
  - SPA
  - Company Websites (SaaS-like websites seems to be a thing)
  - ...
- **State Management**
  With all the tools we have, it comes with a lot of confusion in the community and should have it's own category. It probably makes sense to explain the differences between client-side-state and server-side-state and why you probably shouldn't manage server-side-state manually.
- **Useful how to's**
  Something like "How to move from SASS to Styled Components"

**Reply by markerikson** [reply]
Yeah, I see this as a mixture of "what are the major/possible options out there in a given tools category?", "what problems do those options even solve?", and "when and why should I consider using option A or option B?".

I also like this suggestion ( https://twitter.com/markspolakovs/status/1366302847843844097 ):

> require each tool to have both a “this is what problems this tool solves” as well as a “this is when you should use something else” section? May mitigate the “TOOL X IS OMNIPOTENT” effect.

**Comment by jordyvandomselaar** [comment]
I think we could look a bit at  how awesome lists work:

**State management**

 - Redux
 - MobX
 - ...

**Tooling**

- Create-react-app
- Yeoman
- Next
- ...


and then have something like a checkbox to select them and be able to compare their pro's and cons side-by-side.

If you combine that with tutorials like 

> Let's build a to-do app

Which mixes and matches different libraries so you can see the difference. It could contain Redux and MobX so you can see exactly how different it is in the context of such an app.

Or even

> Adding state to a component

And then showing the difference between hooks, MobX and for example Apollo would be really helpful I think.


As for moderation, we could approach this with flat files. Something like markdown + frontmatter for meta data like tags etc. Authors would then simply make pull requests to add their library and the reviewers could then triage. Markdown would allow authors to add examples and context while still adhering to the styling of this tool so it's a more streamlined experience. Frontmatter would allow authors to add tags like `state management` so you can search for keywords and find libraries that match your query.

Something that would also help is getting respected community members on board. This would _combat_ questions like 

> Why would I believe this website and their best practices when my buddy says I should just _insert footgun here_

**Reply by markerikson** [reply]
At least to start with, I'm picturing something less dynamic.  Like, say, a Docusaurus site with categories.  I'd rather start simpler and focus on getting some meaningful material and traction first, and then worry about building something fancy later.

Similarly, while I can see the potential value in having comparisons of how to build the same app different ways, I'd rather not go that direction for now.  There are after all entire projects like https://github.com/gothinkster/realworld for that already.  I'd rather focus on some descriptive content for now.

**Reply by jordyvandomselaar** [reply]
> a Docusaurus site with categories.

Sounds good to me! I've had a quick look at Docusaurus and it does look like it could work very well.

**Reply by markerikson** [reply]
For reference, all of the Redux lib doc sites are currently running on Docusaurus ( https://redux.js.org , https://react-redux.js.org , https://redux-toolkit.js.org , https://rtk-query-docs.netlify.app ).  For that matter, so is the React TypeScript Cheatsheet ( https://react-typescript-cheatsheet.netlify.app/docs/basic/setup ).

So, it's certainly something _I'm_ familiar with, and it's got a lot of good stuff that's relevant here: Markdown content, syntax highlighting, Table of Contents, etc.  We can always build something fancier later, but in terms of getting an MVP of this idea off the ground, I think it's more than sufficient to get started with.

**Comment by MDanialSaleem** [comment]
I have two opinions on this. 
1. I saw your tweet saying that it will be "slightly" opinionated. I think we should go full opinionated. Or maybe something like, here's the suggested pick for this particular problem (e.g go with redux for statement which is true in most cases I guess) and this suggested pick must be highlighted very strongly. Along with that, we can go like, here are the other options here, and a 1-2 line description of what **specific circumstances may cause you to choose that over the suggested pick** 
2. It maybe possible to create sort of a  UI, which lets you set some configs for your project, consider for example the size of the project, and other stuff, and on the basis of that we spit out the best configs possible.

But anyway, this is such a great initiative. I am sick of having to compare a shit ton of options myself when I stumble upon an issue that I have not solved recently.

**Reply by markerikson** [reply]
I definitely want to keep this limited in scope and goals, at least to start out with.  It's a lot easier to write up a 500-1000 word Markdown file that says "Redux is a state management tool, it helps solve these problems, here's when you might want to use it and here's when you should look at other options", than it is to go create a site that does config customization.

When I say it should be "slightly opinionated", what I mean is that we ought to list the top few options for each category of tools, not list every tool out there that might vaguely fit this category, and focus on "when does it make sense to use this tool" vs "USE TOOL X INSTEAD OF TOOL Y".  So, curated, and listing options, instead of necessarily saying "Use Redux instead of Mobx", etc.

**Reply by andrei9669** [reply]
>  I think we should go full opinionated

I actually kind of agree with this. Reacts biggest problem IMO is that everyone can write it however they want. If there was a way to standardize stuff it would be so much easier to move between projects.

In short, it should be as opinionated as possible but leave enough room for some edge cases if you really must do it some other way.

**Reply by councit** [reply]
Something that I was thinking while reading through this discussion is a prescriptive build tool/plan. Create a questionnaire and compile the tools based on user inputs. Most likely out of scope for this project, but might be another good way to provide value to the users.

**Comment by jordyvandomselaar** [comment]
I'm thinking there should also be clear guidelines to adhere to if someone would like to add content featuring certain libraries. This makes reviewing/regulating easier and prevents getting bombarded with libraries that don't really add anything.

I think it'll be fun if this takes off to look at the side effects. Instead of creating yet another fetching library that's the same, we might actually start seeing different stuff.

**Reply by councit** [reply]
Perhaps something like a scorecard that is used to compare performance/features ect. This would be cool, and something that users can quickly glance at for comparison summaries.

**Comment by CarlMungazi** [comment]
Whilst not directly the goal of this initiative, another great thing I can see coming out of this would be some educational content around the _whys_ of all these tools. As a longer-term goal, being able to clearly articulate the core problem each of these tools are solving will empower developers (especially beginners) and make it easier for them to assess the options. One of the earlier commenters stated a site such as this can help people realise that `react-query` isn't a state management tool, for example. In some cases that misconception is due to a misunderstanding of why state management is a problem in the first place and how this problems manifests itself in the frontend world. 

I know when I first started out, "best-practice" type resources where very useful but their usefulness took on a new level once I began to understand the core problem being solved by my chosen tool.

**Comment by markerikson** [comment]
This is my first time using GH Discussions, and it feels _just_ different enough than Issues to be a bit disorienting :)  like, I'm not sure where this additional comment is going to end up thread-wise.

Awright, so here's my plan at this point.  I've got another RTK docs page I want to finish tomorrow, and then I want to try to create a skeleton for this site for tomorrow.  What I'm thinking of is a Docusaurus v2 site, hosted on Netlify, and I can try throwing together a first stab at a page on Redux to help serve as a starting point for us to flesh out the intended content, tone, and approach for this site.

As additional next steps, I think we need to:

- try to come up with specific answers for some of the questions I listed in the intro so that we can use those as guidelines.
- start coming up with a list of categories and topics that we think it would be valuable to cover here

**Reply by markerikson** [reply]
Aaaaand we're live!

https://react-community-tools-practices-cheatsheet.netlify.app/

Just an empty Docusaurus site atm, but that's a starting point.

Thanks to @AmyShackles for setting up Docusaurus!

**Reply by markerikson** [reply]
And I've followed this up by sketching out the initial content for the "State Management" section, writing a "State Management Overview" page, and adding a page that describes what Redux is, its tradeoffs, and when you might (or might not) want to use it:

- https://react-community-tools-practices-cheatsheet.netlify.app/state-management/overview
- https://react-community-tools-practices-cheatsheet.netlify.app/state-management/redux

We can keep this discussion open, but it might be time to start dealing with things more at the Issues level?

**Comment by zachdtaylor** [comment]
I think for each library/topic it would be nice to include a section of: Here's how you could do this with plain old React. I'm thinking of something like [this](https://dev.to/davidkpiano/no-disabling-a-button-is-not-app-logic-598i) where David shows how to build a state machine with plain reducers, then as a final point shows where the library comes in handy. You can actually get pretty far just using React, and I think because these third party libraries are talked about so much, people feel like they need to use them right off the bat. In my opinion, you should only use a library when doing it the plain React way becomes unmanageable. Of course, some things become unmanageable in plain React much quicker than others, but I think discussing the plain React way is enlightening because it shows you why the library exists in the first place and why you need or don't need it.

**Reply by markerikson** [reply]
Mmm... agree that there should definitely be some sections with examples of various patterns and techniques that involve just React itself.

But part of the point I'm envisioning _is_ to document the ecosystem and when it makes sense to use a given tool. Not that it's _required_ that you use Tool X when you have that need, but rather "if you have Problem Y, Tool X may be useful at this point".

**Reply by zachdtaylor** [reply]
Maybe I can rephrase: I think discussing patterns / techniques in plain React _is_ what helps people understand when it makes sense to use a given tool. The way I see it, the ecosystem only exists _because_ someone was using React, then ran into a situation where React itself wasn't enough, so they created a library. At least for me, understanding _why_ the tools exist makes it clear _when_ I might want to use them.

**Comment by justrag** [comment]
I think https://www.robinwieruch.de/react-libraries may also be an inspiration?

**Comment by andrei9669** [comment]
@markerikson Do we have some contributors info or something?
I would like to contribute but I have no idea where to start since I don't have experience with contributing to OS projects.
Thanks.

**Reply by markerikson** [reply]
Hi!  No formal `CONTRIBUTORS` document atm, because right now there's just this one discussion thread and the proof of concept site I whipped up.

Really, what we need at this point is people to start picking topics, writing pages with useful descriptions and tradeoffs, and filing PRs to add that info.

Happy to discuss specifics if you'd like!

**Comment by dschinkel** [comment]
Why don't we just start reading and sharing and passing articles and books by those who have come before us. Fowler, Beck, and 100s of others who came up with stuff before most of all of us were even born....all of which we can use to drive "good practices" in any kind of code we're working in.  How about we start listing that kind of stuff in FB docs ay?

Start with "Humble Views / Humble Object".  https://martinfowler.com/bliki/HumbleObject.html Wisdom for the ages.  Figure out how to apply that in React...
4 Simple Rules of Design https://martinfowler.com/bliki/BeckDesignRules.html
Test Desiderata https://medium.com/@kentbeck_7670/test-desiderata-94150638a4b3
Clean Code Practices
SOLID
...the list goes on and on

how can we model good practices like these in React, TS, whatever code we're in.  This is really what's missing.  Read what's already out there, been there for years.

**Comment by mlynchdev** [comment]
If we ask very nicely maybe @jamonholmgren (maintainer of MobX State Tree) could help with the MobX portion and include some portions on MST as well.
